multiple windows

If you build several windows:
```cpp
#include <vector>
#include "include/glfwpp/glfwpp.h"
static void keyboard_callback(glfw::Window& W, glfw::KeyCode K, int, glfw::KeyState S, glfw::ModifierKeyBit)
{
  if(K == glfw::KeyCode::Escape && S == glfw::KeyState::Release) W.setShouldClose(true);
}

int main()
{
  std::vector<glfw::Window> DSE {};

  [[maybe_unused]]auto GLFW = glfw::init();

  DSE.emplace_back( 640, 480, "GLFWPP N_1" );
  DSE.back().keyEvent.setCallback(keyboard_callback);

  DSE.emplace_back( 640, 480, "GLFWPP N_2" );
  DSE.back().keyEvent.setCallback(keyboard_callback);

  DSE.emplace_back( 640, 480, "GLFWPP N_3" );
  DSE.back().keyEvent.setCallback(keyboard_callback);

  while (!DSE.empty())
  {
    for(auto& W: DSE)
    {
      glfw::makeContextCurrent(W);
      glfw::pollEvents();

      if(W.shouldClose())
      {
        auto it = std::find_if(DSE.begin(), DSE.end(), [&W](const auto& cW){ return cW == W; });
        if(it != DSE.end()) DSE.erase(it);
      }
    }
  }
}
```
then the event handler only connects to the latter. Maybe I'm doing wrong?

Strange, I can't compile this branch (windows10, msys2 - mingw64 - clang64 -ninja):

```bash
In file included from D:\dev\cpp\muti-win-cpp\main.cpp:7:
In file included from D:\dev\cpp\muti-win-cpp/include/glfwpp/glfwpp.h:13:
D:\dev\cpp\muti-win-cpp/include/glfwpp/window.h:193:20: error: no matching constructor for initialization of
      'glfw::Cursor'
            return Cursor{glfwCreateCursor(&image_, xHot_, yHot_)};
                   ^     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
D:\dev\cpp\muti-win-cpp/include/glfwpp/window.h:172:11: note: candidate constructor (the implicit copy constructor) not
      viable: cannot convert argument of incomplete type 'GLFWcursor *' to 'const glfw::Cursor' for 1st argument
    class Cursor : public detail::OwningPtr<GLFWcursor>
```

But if you say the example works, then fine. Let's close the issue.

--- a/examples/CMakeLists.txt
+++ b/examples/CMakeLists.txt
@@ -21,6 +21,10 @@ if (NOT (CMAKE_SYSTEM_NAME STREQUAL "Emscripten"))
 	target_link_libraries(glfwpp_example_imgui_integration PRIVATE libglew_static GLFWPP Imgui)
 	target_compile_definitions(glfwpp_example_imgui_integration PRIVATE GLEW_STATIC)
 
+	add_executable(glfwpp_example_multiple_windows multiple_windows.cpp)
+	target_link_libraries(glfwpp_example_multiple_windows PRIVATE libglew_static GLFWPP)
+	target_compile_definitions(glfwpp_example_multiple_windows PRIVATE GLEW_STATIC)
+
 	add_executable(glfwpp_test_multiple_tus multiple_tus1.cpp multiple_tus2.cpp)
 	target_link_libraries(glfwpp_test_multiple_tus PRIVATE libglew_static GLFWPP)
 	target_compile_definitions(glfwpp_test_multiple_tus PRIVATE GLEW_STATIC)
@@ -34,6 +38,7 @@ if (NOT (CMAKE_SYSTEM_NAME STREQUAL "Emscripten"))
 		glfwpp_example_events
 		glfwpp_example_window_attributes
 		glfwpp_example_imgui_integration
+		glfwpp_example_multiple_windows
 		glfwpp_test_multiple_tus
 		glfwpp_test_type_traits
 	)

--- a/examples/multiple_windows.cpp
+++ b/examples/multiple_windows.cpp
@@ -1,43 +1,37 @@
 #include <glfwpp/glfwpp.h>
+#include <string>
 #include <vector>
 
-static void keyboard_callback(glfw::Window& W, glfw::KeyCode K, int, glfw::KeyState S, glfw::ModifierKeyBit)
+// based on https://github.com/janekb04/glfwpp/issues/136
+
+static void keyboard_callback(glfw::Window& wnd, glfw::KeyCode K, int, glfw::KeyState S, glfw::ModifierKeyBit)
 {
     if(K == glfw::KeyCode::Escape && S == glfw::KeyState::Release)
-        W.setShouldClose(true);
+    {
+        wnd.setShouldClose(true);
+    }
 }
 
 int main()
 {
-    std::vector<glfw::Window> DSE{};
+    std::vector<glfw::Window> v{};
 
     [[maybe_unused]] auto GLFW = glfw::init();
 
-    //DSE.reserve(10);
-
-    DSE.emplace_back(640, 480, "GLFWPP N_1");
-    DSE.back().keyEvent.setCallback(keyboard_callback);
-
-    DSE.emplace_back(640, 480, "GLFWPP N_2");
-    DSE.back().keyEvent.setCallback(keyboard_callback);
-
-    DSE.emplace_back(640, 480, "GLFWPP N_3");
-    DSE.back().keyEvent.setCallback(keyboard_callback);
-
-    while(!DSE.empty())
+    for(int windows = 0; windows < 11; ++windows)
     {
-        for(auto& W : DSE)
-        {
-            glfw::makeContextCurrent(W);
-            glfw::pollEvents();
+        v.emplace_back(640, 480, ("Window " + std::to_string(windows)).c_str());
+        v.back().keyEvent.setCallback(keyboard_callback);
+    }
 
+    while(!v.empty())
+    {
+        glfw::pollEvents();
 
-            if(W.shouldClose())
-            {
-                auto it = std::find_if(DSE.begin(), DSE.end(), [&W](const auto& cW) { return cW == W; });
-                if(it != DSE.end())
-                    DSE.erase(it);
-            }
-        }
+        v.erase(
+                std::remove_if(v.begin(), v.end(), [](const glfw::Window& wnd) {
+                    return wnd.shouldClose();
+                }),
+                v.end());
     }
 }

--- a/examples/multiple_windows.cpp
+++ b/examples/multiple_windows.cpp
@@ -1,0 +1,43 @@
+#include <glfwpp/glfwpp.h>
+#include <vector>
+
+static void keyboard_callback(glfw::Window& W, glfw::KeyCode K, int, glfw::KeyState S, glfw::ModifierKeyBit)
+{
+    if(K == glfw::KeyCode::Escape && S == glfw::KeyState::Release)
+        W.setShouldClose(true);
+}
+
+int main()
+{
+    std::vector<glfw::Window> DSE{};
+
+    [[maybe_unused]] auto GLFW = glfw::init();
+
+    //DSE.reserve(10);
+
+    DSE.emplace_back(640, 480, "GLFWPP N_1");
+    DSE.back().keyEvent.setCallback(keyboard_callback);
+
+    DSE.emplace_back(640, 480, "GLFWPP N_2");
+    DSE.back().keyEvent.setCallback(keyboard_callback);
+
+    DSE.emplace_back(640, 480, "GLFWPP N_3");
+    DSE.back().keyEvent.setCallback(keyboard_callback);
+
+    while(!DSE.empty())
+    {
+        for(auto& W : DSE)
+        {
+            glfw::makeContextCurrent(W);
+            glfw::pollEvents();
+
+
+            if(W.shouldClose())
+            {
+                auto it = std::find_if(DSE.begin(), DSE.end(), [&W](const auto& cW) { return cW == W; });
+                if(it != DSE.end())
+                    DSE.erase(it);
+            }
+        }
+    }
+}

--- a/include/glfwpp/helper.h
+++ b/include/glfwpp/helper.h
@@ -21,4 +21,51 @@
         return static_cast<std::underlying_type_t<Enum>>(lhs) ^ static_cast<std::underlying_type_t<Enum>>(rhs); \
     }
 
+namespace glfw
+{
+    namespace detail
+    {
+        template<typename T>
+        class OwningPtr
+        {
+        private:
+            T* _ptr;
+
+        public:
+            OwningPtr(const OwningPtr&) = delete;
+            OwningPtr& operator=(const OwningPtr&) = delete;
+
+            constexpr OwningPtr(std::nullptr_t = nullptr) noexcept :
+                _ptr{}
+            {
+            }
+
+            constexpr OwningPtr(T* ptr_) :
+                _ptr{ptr_}
+            {
+            }
+
+            constexpr OwningPtr(OwningPtr&& other) noexcept :
+                _ptr{std::exchange(other._ptr, nullptr)}
+            {
+            }
+
+            constexpr OwningPtr& operator=(OwningPtr&& other) noexcept
+            {
+                _ptr = std::exchange(other._ptr, nullptr);
+            }
+
+            [[nodiscard]] constexpr operator bool() const noexcept
+            {
+                return static_cast<bool>(_ptr);
+            }
+
+            [[nodiscard]] constexpr operator T*() const noexcept
+            {
+                return _ptr;
+            }
+        };
+    }  // namespace detail
+}  // namespace glfw
+
 #endif  //GLFWPP_HELPER_H

--- a/include/glfwpp/helper.h
+++ b/include/glfwpp/helper.h
@@ -53,6 +53,7 @@ namespace glfw
             constexpr OwningPtr& operator=(OwningPtr&& other) noexcept
             {
                 _ptr = std::exchange(other._ptr, nullptr);
+                return *this;
             }
 
             [[nodiscard]] constexpr explicit operator bool() const noexcept

--- a/include/glfwpp/helper.h
+++ b/include/glfwpp/helper.h
@@ -55,7 +55,7 @@ namespace glfw
                 _ptr = std::exchange(other._ptr, nullptr);
             }
 
-            [[nodiscard]] constexpr operator bool() const noexcept
+            [[nodiscard]] constexpr explicit operator bool() const noexcept
             {
                 return static_cast<bool>(_ptr);
             }

--- a/include/glfwpp/monitor.h
+++ b/include/glfwpp/monitor.h
@@ -54,7 +54,6 @@ namespace glfw
         {
             return Monitor{glfwGetPrimaryMonitor()};
         }
-        friend class Window;
 
     public:
         [[nodiscard]] std::tuple<int, int> getPos() const

--- a/include/glfwpp/window.h
+++ b/include/glfwpp/window.h
@@ -425,7 +425,31 @@ namespace glfw
     class Window
     {
     private:
-        detail::OwningPtr<GLFWwindow> _handle;
+        class HandleContainer : public detail::OwningPtr<GLFWwindow>
+        {
+        public:
+            using detail::OwningPtr<GLFWwindow>::OwningPtr;
+
+            HandleContainer(HandleContainer&& other) :
+                detail::OwningPtr<GLFWwindow>{std::move(other)}
+            {
+                // NOTE: We use the fact that _handle is the first member of
+                // class Window, which means that its address `this` is equal
+                // to the address `this` of its containing Window object.
+                // This is done to prevent HandleContainer from needing to have
+                // a pointer to the enclosing Window class object.
+                _setPointerFromHandle(static_cast<GLFWwindow*>(*this), reinterpret_cast<Window*>(this));
+            }
+
+            HandleContainer& operator=(HandleContainer&& other)
+            {
+                static_cast<detail::OwningPtr<GLFWwindow>&>(*this) = std::move(other);
+                // NOTE: as above
+                _setPointerFromHandle(static_cast<GLFWwindow*>(*this), reinterpret_cast<Window*>(this));
+
+                return *this;
+            }
+        } _handle;
         detail::OwningPtr<void> _userPtr;
 
     public:
@@ -616,28 +640,9 @@ namespace glfw
 
         Window& operator=(const Window&) = delete;
 
-        Window(Window&& other) noexcept :
-            _handle{std::exchange(other._handle, nullptr)},
-            _userPtr{std::exchange(other._userPtr, nullptr)}
-        {
-            if(_handle)
-            {
-                _setPointerFromHandle(_handle, this);
-            }
-        }
+        Window(Window&& other) noexcept = default;
 
-        Window& operator=(Window&& other) noexcept
-        {
-            _handle = std::exchange(other._handle, nullptr);
-            _userPtr = std::exchange(other._userPtr, nullptr);
-
-            if(_handle)
-            {
-                _setPointerFromHandle(_handle, this);
-            }
-
-            return *this;
-        }
+        Window& operator=(Window&& other) = default;
 
         ~Window()
         {

--- a/include/glfwpp/window.h
+++ b/include/glfwpp/window.h
@@ -172,14 +172,17 @@ namespace glfw
     class Cursor : public detail::OwningPtr<GLFWcursor>
     {
     public:
-        Cursor& operator=(Cursor&& other)
+        Cursor() noexcept = default;
+        Cursor(Cursor&&) noexcept = default;
+
+        Cursor& operator=(Cursor&& other) noexcept
         {
             glfwDestroyCursor(static_cast<GLFWcursor*>(*this));
             static_cast<detail::OwningPtr<GLFWcursor>&>(*this) = std::move(other);
             return *this;
         }
 
-        ~Cursor()
+        ~Cursor() noexcept
         {
             glfwDestroyCursor(static_cast<GLFWcursor*>(*this));
         }
@@ -437,7 +440,7 @@ namespace glfw
         public:
             using detail::OwningPtr<GLFWwindow>::OwningPtr;
 
-            HandleContainer(HandleContainer&& other) :
+            HandleContainer(HandleContainer&& other) noexcept :
                 detail::OwningPtr<GLFWwindow>{std::move(other)}
             {
                 // NOTE: We use the fact that _handle is the first member of
@@ -449,7 +452,7 @@ namespace glfw
                     _setPointerFromHandle(static_cast<GLFWwindow*>(*this), reinterpret_cast<Window*>(this));
             }
 
-            HandleContainer& operator=(HandleContainer&& other)
+            HandleContainer& operator=(HandleContainer&& other) noexcept
             {
                 glfwDestroyWindow(static_cast<GLFWwindow*>(*this));
                 static_cast<detail::OwningPtr<GLFWwindow>&>(*this) = std::move(other);
@@ -460,7 +463,7 @@ namespace glfw
                 return *this;
             }
 
-            ~HandleContainer()
+            ~HandleContainer() noexcept
             {
                 glfwDestroyWindow(static_cast<GLFWwindow*>(*this));
             }
@@ -657,7 +660,7 @@ namespace glfw
 
         Window(Window&& other) noexcept = default;
 
-        Window& operator=(Window&& other) = default;
+        Window& operator=(Window&& other) noexcept = default;
 
         //Retains ownership
         operator GLFWwindow*() const

--- a/include/glfwpp/window.h
+++ b/include/glfwpp/window.h
@@ -172,6 +172,8 @@ namespace glfw
     class Cursor : public detail::OwningPtr<GLFWcursor>
     {
     public:
+        using detail::OwningPtr<GLFWcursor>::OwningPtr;
+
         Cursor() noexcept = default;
         Cursor(Cursor&&) noexcept = default;
 

--- a/include/glfwpp/window.h
+++ b/include/glfwpp/window.h
@@ -172,6 +172,13 @@ namespace glfw
     class Cursor : public detail::OwningPtr<GLFWcursor>
     {
     public:
+        Cursor& operator=(Cursor&& other)
+        {
+            glfwDestroyCursor(static_cast<GLFWcursor*>(*this));
+            static_cast<detail::OwningPtr<GLFWcursor>&>(*this) = std::move(other);
+            return *this;
+        }
+
         ~Cursor()
         {
             glfwDestroyCursor(static_cast<GLFWcursor*>(*this));
@@ -444,12 +451,18 @@ namespace glfw
 
             HandleContainer& operator=(HandleContainer&& other)
             {
+                glfwDestroyWindow(static_cast<GLFWwindow*>(*this));
                 static_cast<detail::OwningPtr<GLFWwindow>&>(*this) = std::move(other);
                 // NOTE: as above
                 if(static_cast<GLFWwindow*>(*this))
                     _setPointerFromHandle(static_cast<GLFWwindow*>(*this), reinterpret_cast<Window*>(this));
 
                 return *this;
+            }
+
+            ~HandleContainer()
+            {
+                glfwDestroyWindow(static_cast<GLFWwindow*>(*this));
             }
         } _handle;
         detail::OwningPtr<void> _userPtr;
@@ -645,11 +658,6 @@ namespace glfw
         Window(Window&& other) noexcept = default;
 
         Window& operator=(Window&& other) = default;
-
-        ~Window()
-        {
-            glfwDestroyWindow(_handle);
-        }
 
         //Retains ownership
         operator GLFWwindow*() const

--- a/include/glfwpp/window.h
+++ b/include/glfwpp/window.h
@@ -169,52 +169,13 @@ namespace glfw
     };
     using Image = GLFWimage;
 
-    class Cursor
+    class Cursor : public detail::OwningPtr<GLFWcursor>
     {
-    private:
-        GLFWcursor* _handle;
-
-        friend class Window;
-
     public:
-        Cursor() :
-            Cursor{nullptr}
-        {
-        }
-
-        // Takes ownership
-        explicit Cursor(GLFWcursor* handle_) :
-            _handle{handle_}
-        {
-        }
-
-        Cursor(const Cursor&) = delete;
-
-        Cursor& operator=(const Cursor&) = delete;
-
-        Cursor(Cursor&& other) noexcept :
-            _handle{std::exchange(other._handle, nullptr)}
-        {
-        }
-
-        Cursor& operator=(Cursor&& other) noexcept
-        {
-            _handle = std::exchange(other._handle, nullptr);
-            return *this;
-        }
-
         ~Cursor()
         {
-            glfwDestroyCursor(_handle);
+            glfwDestroyCursor(static_cast<GLFWcursor*>(*this));
         }
-
-        // Retains ownership
-        explicit operator GLFWcursor*() const
-        {
-            return _handle;
-        }
-
-        explicit operator bool() const = delete;
 
     public:
         [[nodiscard]] static Cursor createCursor(const Image& image_, int xHot_, int yHot_)
@@ -464,8 +425,8 @@ namespace glfw
     class Window
     {
     private:
-        GLFWwindow* _handle;
-        void* _userPtr;
+        detail::OwningPtr<GLFWwindow> _handle;
+        detail::OwningPtr<void> _userPtr;
 
     public:
         Event<Window&, int, int> posEvent;
@@ -598,16 +559,16 @@ namespace glfw
         }
 
     public:
-        explicit Window(std::nullptr_t = nullptr) :
-            _handle{nullptr},
-            _userPtr{nullptr}
+        explicit Window() noexcept = default;
+        explicit Window(std::nullptr_t) noexcept :
+            Window{}
         {
         }
 
         //Takes ownership
         explicit Window(GLFWwindow* handle_) :
             _handle{handle_},
-            _userPtr{nullptr}
+            _userPtr{}
         {
             if(_handle)
             {
@@ -646,8 +607,8 @@ namespace glfw
                     width_,
                     height_,
                     title_,
-                    monitor_ ? monitor_->_handle : nullptr,
-                    share_ ? share_->_handle : nullptr)}
+                    monitor_ ? static_cast<GLFWmonitor*>(*monitor_) : nullptr,
+                    share_ ? static_cast<GLFWwindow*>(share_->_handle) : nullptr)}
         {
         }
 
@@ -819,7 +780,7 @@ namespace glfw
 
         void setMonitor(Monitor monitor_, int xPos_, int yPos_, int width_, int height_, int refreshRate_)
         {
-            glfwSetWindowMonitor(_handle, monitor_._handle, xPos_, yPos_, width_, height_, refreshRate_);
+            glfwSetWindowMonitor(_handle, static_cast<GLFWmonitor*>(monitor_), xPos_, yPos_, width_, height_, refreshRate_);
         }
 
         [[nodiscard]] bool getAttribFocused() const
@@ -1063,7 +1024,7 @@ namespace glfw
 
         void setCursor(const Cursor& cursor_)
         {
-            glfwSetCursor(_handle, cursor_._handle);
+            glfwSetCursor(_handle, static_cast<GLFWcursor*>(cursor_));
         }
 
 #if defined(VK_VERSION_1_0)

--- a/include/glfwpp/window.h
+++ b/include/glfwpp/window.h
@@ -438,14 +438,16 @@ namespace glfw
                 // to the address `this` of its containing Window object.
                 // This is done to prevent HandleContainer from needing to have
                 // a pointer to the enclosing Window class object.
-                _setPointerFromHandle(static_cast<GLFWwindow*>(*this), reinterpret_cast<Window*>(this));
+                if(static_cast<GLFWwindow*>(*this))
+                    _setPointerFromHandle(static_cast<GLFWwindow*>(*this), reinterpret_cast<Window*>(this));
             }
 
             HandleContainer& operator=(HandleContainer&& other)
             {
                 static_cast<detail::OwningPtr<GLFWwindow>&>(*this) = std::move(other);
                 // NOTE: as above
-                _setPointerFromHandle(static_cast<GLFWwindow*>(*this), reinterpret_cast<Window*>(this));
+                if(static_cast<GLFWwindow*>(*this))
+                    _setPointerFromHandle(static_cast<GLFWwindow*>(*this), reinterpret_cast<Window*>(this));
 
                 return *this;
             }

